<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content='PREGO implements regression algorithms for finding motifs in DNA. It contains algorithms which were described at: "Extensive low-affinity transcriptional interactions in the yeast genome" &lt;doi:10.1101/gr.5113606&gt;.'>
<title>PWM Regression Optimizer • prego</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="PWM Regression Optimizer">
<meta property="og:description" content='PREGO implements regression algorithms for finding motifs in DNA. It contains algorithms which were described at: "Extensive low-affinity transcriptional interactions in the yeast genome" &lt;doi:10.1101/gr.5113606&gt;.'>
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">prego</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="prego">prego<a class="anchor" aria-label="anchor" href="#prego"></a>
</h1></div>
<!-- badges: start -->
<p>The <code>prego</code> package implements simple regression algorithms for finding motifs in DNA. You can either use it to find motif which are discriminating between two or more clusters of DNA sequences, or for generting motifs from one or more continuous variables.</p>
<p>The PREGO algorithm is described <a href="https://doi.org/10.1101%2Fgr.5113606" class="external-link">here</a>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of prego like so:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remotes/man/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"tanaylab/prego"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">prego</span><span class="op">)</span></span></code></pre></div>
<p>For a set of continuous variables:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/regress_pwm.html">regress_pwm</a></span><span class="op">(</span><span class="va">sequences_example</span>, <span class="va">response_mat_example</span><span class="op">)</span></span>
<span><span class="co">#&gt; ℹ Number of response variables: 5</span></span>
<span><span class="co">#&gt; ℹ Screening for kmers in order to initialize regression</span></span>
<span><span class="co">#&gt; ℹ Number of response variables: 5</span></span>
<span><span class="co">#&gt; ℹ Screening kmers of length 8, from position 0 to position 300</span></span>
<span><span class="co">#&gt; ℹ minimal correlation: 0.08, minimal number of occurrences: 50</span></span>
<span><span class="co">#&gt; ✔ Found 2138 kmers in 1000 sequences.</span></span>
<span><span class="co">#&gt; ℹ Motif is shorter than 15, extending with wildcards</span></span>
<span><span class="co">#&gt; ℹ Initializing regression with the following motif: "***TTTACAAC****"</span></span>
<span><span class="co">#&gt; ℹ Running regression</span></span>
<span><span class="co">#&gt; • Motif length: 15</span></span>
<span><span class="co">#&gt; • Bidirectional: TRUE</span></span>
<span><span class="co">#&gt; • Spat min: 0</span></span>
<span><span class="co">#&gt; • Spat max: 300</span></span>
<span><span class="co">#&gt; • Spat bin: 50</span></span>
<span><span class="co">#&gt; • Improve epsilon: 0.0001</span></span>
<span><span class="co">#&gt; • Min nuc prob: 0.001</span></span>
<span><span class="co">#&gt; • Uniform prior: 0.05</span></span>
<span><span class="co">#&gt; • Score metric: "r2"</span></span>
<span><span class="co">#&gt; • Seed: 60427</span></span>
<span><span class="co">#&gt; ✔ Finished running regression. Consensus: "TTACRAC"</span></span>
<span><span class="co">#&gt; ✔ R^2: 0.01, 0.0054, 0.0373, 0.0002, and 0.0087</span></span>
<span><span class="fu"><a href="reference/plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")`</span></span>
<span><span class="co">#&gt; instead.</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-3-1.png" width="100%"></p>
<p>For binary response:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_binary</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/regress_pwm.html">regress_pwm</a></span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">cluster_mat_example</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; ℹ Number of response variables: 1</span></span>
<span><span class="co">#&gt; ℹ Screening for kmers in order to initialize regression</span></span>
<span><span class="co">#&gt; ℹ Number of response variables: 1</span></span>
<span><span class="co">#&gt; ℹ Screening kmers of length 8, from position 0 to position 300</span></span>
<span><span class="co">#&gt; ℹ minimal correlation: 0.08, minimal number of occurrences: 50</span></span>
<span><span class="co">#&gt; ✔ Found 198 kmers in 2359 sequences.</span></span>
<span><span class="co">#&gt; ℹ Motif is shorter than 15, extending with wildcards</span></span>
<span><span class="co">#&gt; ℹ Initializing regression with the following motif: "***TAATCATT****"</span></span>
<span><span class="co">#&gt; ℹ Running regression</span></span>
<span><span class="co">#&gt; • Motif length: 15</span></span>
<span><span class="co">#&gt; • Bidirectional: TRUE</span></span>
<span><span class="co">#&gt; • Spat min: 0</span></span>
<span><span class="co">#&gt; • Spat max: 300</span></span>
<span><span class="co">#&gt; • Spat bin: 50</span></span>
<span><span class="co">#&gt; • Improve epsilon: 0.0001</span></span>
<span><span class="co">#&gt; • Min nuc prob: 0.001</span></span>
<span><span class="co">#&gt; • Uniform prior: 0.05</span></span>
<span><span class="co">#&gt; • Score metric: "r2"</span></span>
<span><span class="co">#&gt; • Seed: 60427</span></span>
<span><span class="co">#&gt; ✔ Finished running regression. Consensus: "T*A***W*T"</span></span>
<span><span class="co">#&gt; ✔ KS test D: 0.8507, p-value: 0</span></span>
<span><span class="fu"><a href="reference/plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res_binary</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")`</span></span>
<span><span class="co">#&gt; instead.</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-4-1.png" width="100%"></p>
<p>For clusters of sequences:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/regress_pwm.clusters.html">regress_pwm.clusters</a></span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">clusters_example</span><span class="op">)</span></span>
<span><span class="co">#&gt; ℹ Using two-phase optimization</span></span>
<span><span class="co">#&gt; ℹ Running regression for 2359 clusters</span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; # A tibble: 5 x 5</span></span>
<span><span class="co">#&gt;   cluster consensus      ks_D        r2      seed_motif</span></span>
<span><span class="co">#&gt; 1    c100   AT***TC 0.6998806 0.4004041 ***ATCCATCA****</span></span>
<span><span class="co">#&gt; 2    c111   Y*RTAAA 0.8492008 0.5081351 ****CATAAA*****</span></span>
<span><span class="co">#&gt; 3     c29 R*W**KT*A 0.8514453 0.5724329 ***AAT*ATTAA***</span></span>
<span><span class="co">#&gt; 4      c5      &lt;NA&gt; 0.5543292 0.2086212 ***AAT*ATTAA***</span></span>
<span><span class="co">#&gt; 5      c6      WATC 0.6016048 0.2819655 ***TCTTATCT****</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing prego</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Aviezer Lifshitz <br><small class="roles"> Author, maintainer </small>  </li>
<li>Amos Tanay <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://CRAN.R-project.org/package=prego" class="external-link"><img src="https://www.r-pkg.org/badges/version/prego" alt="CRAN status"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Aviezer Lifshitz, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
