<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>PWM Regression Optimizer • prego</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="PWM Regression Optimizer">
<meta name="description" content='PREGO implements regression algorithms for finding motifs in DNA. It contains algorithms which were described at: "Extensive low-affinity transcriptional interactions in the yeast genome" &lt;doi:10.1101/gr.5113606&gt;.'>
<meta property="og:description" content='PREGO implements regression algorithms for finding motifs in DNA. It contains algorithms which were described at: "Extensive low-affinity transcriptional interactions in the yeast genome" &lt;doi:10.1101/gr.5113606&gt;.'>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">prego</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="prego">prego<a class="anchor" aria-label="anchor" href="#prego"></a>
</h1></div>
<!-- badges: start -->

<p>The <code>prego</code> package implements simple regression algorithms for finding motifs in DNA. You can either use it to find motif which are discriminating between two or more clusters of DNA sequences, or for generating motifs from one or more continuous variables.</p>
<p>The PREGO algorithm is described <a href="https://doi.org/10.1101%2Fgr.5113606" class="external-link">here</a>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of prego like so:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"tanaylab/prego"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">prego</span><span class="op">)</span></span>
<span><span class="co">#&gt; ℹ Parallelization enabled. Using 77 threads. </span></span></code></pre></div>
<p>For a set of continuous variables:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/regress_pwm.html">regress_pwm</a></span><span class="op">(</span><span class="va">sequences_example</span>, <span class="va">response_mat_example</span><span class="op">)</span></span>
<span><span class="co">#&gt; ℹ Using "r2" as the final metric</span></span>
<span><span class="co">#&gt; ℹ Number of response variables: 5</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Generate candidate kmers</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Regress each candidate kmer</span></span>
<span><span class="co">#&gt; ℹ Running regression on 10 candidate kmers</span></span>
<span><span class="co">#&gt; • Bidirectional: TRUE</span></span>
<span><span class="co">#&gt; • Spat bin size: 40</span></span>
<span><span class="co">#&gt; • Number of spatial bins 7</span></span>
<span><span class="co">#&gt; • Length of sequence: 280</span></span>
<span><span class="co">#&gt; • Min gap: 0</span></span>
<span><span class="co">#&gt; • Max gap: 1</span></span>
<span><span class="co">#&gt; • Kmer length: 8</span></span>
<span><span class="co">#&gt; • Improve epsilon: 1e-04</span></span>
<span><span class="co">#&gt; • Min nuc prob: 0.001</span></span>
<span><span class="co">#&gt; • Uniform prior: 0.05</span></span>
<span><span class="co">#&gt; • Score metric: "r2"</span></span>
<span><span class="co">#&gt; • Seed: 60427</span></span>
<span><span class="co">#&gt; • kmer: "TTTACAAC", score (r2): 0.0126890005368537</span></span>
<span><span class="co">#&gt; • kmer: "TA*CTGCTG", score (r2): 0.00334159650449452</span></span>
<span><span class="co">#&gt; • kmer: "AACCATCT", score (r2): 0.0643097230120355</span></span>
<span><span class="co">#&gt; • kmer: "AGA*AAGAA", score (r2): 0.060384757659326</span></span>
<span><span class="co">#&gt; • kmer: "TTATC*TCT", score (r2): 0.0657310811484614</span></span>
<span><span class="co">#&gt; • kmer: "TCACCTTG", score (r2): 0.0623381197705449</span></span>
<span><span class="co">#&gt; • kmer: "CTCTGCAG", score (r2): 0.0677598761958694</span></span>
<span><span class="co">#&gt; • kmer: "TAGCTGTC", score (r2): 0.0640768115164421</span></span>
<span><span class="co">#&gt; • kmer: "GACAAT*CT", score (r2): 0.0607251519441225</span></span>
<span><span class="co">#&gt; • kmer: "TCTTATCT", score (r2): 0.0628163788281552</span></span>
<span><span class="co">#&gt; ℹ Best match in the database: "JASPAR.ct", cor: 0.667</span></span>
<span><span class="co">#&gt; ℹ Best motif: "***CTCTGCAG****", score (r2): 0.0677598761958694</span></span>
<span><span class="fu"><a href="reference/plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: The `&lt;scale&gt;` argument of `guides()` cannot be `FALSE`. Use "none" instead as</span></span>
<span><span class="co">#&gt; of ggplot2 3.3.4.</span></span>
<span><span class="co">#&gt; ℹ The deprecated feature was likely used in the ggseqlogo package.</span></span>
<span><span class="co">#&gt;   Please report the issue at &lt;https://github.com/omarwagih/ggseqlogo/issues&gt;.</span></span>
<span><span class="co">#&gt; This warning is displayed once every 8 hours.</span></span>
<span><span class="co">#&gt; Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span>
<span><span class="co">#&gt; generated.</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-3-1.png" width="100%"></p>
<p>For binary response:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_binary</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/regress_pwm.html">regress_pwm</a></span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">cluster_mat_example</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; ℹ Using "ks" as the final metric</span></span>
<span><span class="co">#&gt; ℹ Number of response variables: 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Generate candidate kmers</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Regress each candidate kmer</span></span>
<span><span class="co">#&gt; ℹ Running regression on 10 candidate kmers</span></span>
<span><span class="co">#&gt; • Bidirectional: TRUE</span></span>
<span><span class="co">#&gt; • Spat bin size: 40</span></span>
<span><span class="co">#&gt; • Number of spatial bins 7</span></span>
<span><span class="co">#&gt; • Length of sequence: 280</span></span>
<span><span class="co">#&gt; • Min gap: 0</span></span>
<span><span class="co">#&gt; • Max gap: 1</span></span>
<span><span class="co">#&gt; • Kmer length: 8</span></span>
<span><span class="co">#&gt; • Improve epsilon: 1e-04</span></span>
<span><span class="co">#&gt; • Min nuc prob: 0.001</span></span>
<span><span class="co">#&gt; • Uniform prior: 0.05</span></span>
<span><span class="co">#&gt; • Score metric: "r2"</span></span>
<span><span class="co">#&gt; • Seed: 60427</span></span>
<span><span class="co">#&gt; • kmer: "TTAAT*ATT", score (ks): 0.854878919705105</span></span>
<span><span class="co">#&gt; • kmer: "AAT*ATTAA", score (ks): 0.854878919705105</span></span>
<span><span class="co">#&gt; • kmer: "TAAT*ATTA", score (ks): 0.843301065576269</span></span>
<span><span class="co">#&gt; • kmer: "GTTAAT*AT", score (ks): 0.844482257088238</span></span>
<span><span class="co">#&gt; • kmer: "AT*ATTAAC", score (ks): 0.844482257088238</span></span>
<span><span class="co">#&gt; • kmer: "AA*CATTAA", score (ks): 0.84950858078479</span></span>
<span><span class="co">#&gt; • kmer: "TTAA*CATT", score (ks): 0.849192870611084</span></span>
<span><span class="co">#&gt; • kmer: "TAATCATT", score (ks): 0.841995767306361</span></span>
<span><span class="co">#&gt; • kmer: "TTA*TCATT", score (ks): 0.845127829960678</span></span>
<span><span class="co">#&gt; • kmer: "TAA*CATTA", score (ks): 0.833248418183164</span></span>
<span><span class="co">#&gt; ℹ Best match in the database: "JOLMA.HNF1B_di_full_1", cor: 0.8</span></span>
<span><span class="co">#&gt; ✔ "JOLMA.HNF1B_di_full_1" KS test D: 0.8512, p-value: 0</span></span>
<span><span class="co">#&gt; ℹ Best motif: "***TTAAT*ATT***", score (ks): 0.854878919705105</span></span>
<span><span class="fu"><a href="reference/plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res_binary</span><span class="op">)</span></span></code></pre></div>
<p><img src="reference/figures/README-unnamed-chunk-4-1.png" width="100%"></p>
<p>For clusters of sequences:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/regress_pwm.clusters.html">regress_pwm.clusters</a></span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">clusters_example</span><span class="op">)</span></span>
<span><span class="co">#&gt; ℹ Using sampled optimization</span></span>
<span><span class="co">#&gt; ℹ Running regression for 5 clusters</span></span>
<span><span class="co">#&gt; ℹ Matching with motif databases</span></span>
<span><span class="va">res</span><span class="op">$</span><span class="va">stats</span></span>
<span><span class="co">#&gt; # A tibble: 5 × 7</span></span>
<span><span class="co">#&gt;   cluster consensus  ks_D    r2 seed_motif      db_match                 db_ma…¹</span></span>
<span><span class="co">#&gt;   &lt;chr&gt;   &lt;chr&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;chr&gt;           &lt;chr&gt;                      &lt;dbl&gt;</span></span>
<span><span class="co">#&gt; 1 c100    GA***AT   0.668 0.326 ***TGATGG*TG*** HOCOMOCO.PDX1_HUMAN.H11…   0.786</span></span>
<span><span class="co">#&gt; 2 c111    Y**TAAA   0.836 0.514 ***GC*ATAAAA*** JOLMA.CDX2_mono_DBD        0.889</span></span>
<span><span class="co">#&gt; 3 c29     T*A***W*Y 0.830 0.552 ***TTAA*CATT*** JOLMA.HNF1B_di_full_1      0.745</span></span>
<span><span class="co">#&gt; 4 c5      &lt;NA&gt;      0.569 0.194 ***TGAT*GATG*** JASPAR.CG15696-RA          0.725</span></span>
<span><span class="co">#&gt; 5 c6      ATC       0.606 0.262 ***TCTTATCT**** HOMER.Gata2                0.864</span></span>
<span><span class="co">#&gt; # … with abbreviated variable name ¹​db_match_cor</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing prego</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Aviezer Lifshitz <br><small class="roles"> Author, maintainer </small>  </li>
<li>Amos Tanay <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://lifecycle.r-lib.org/articles/stages.html#experimental" class="external-link"><img src="https://img.shields.io/badge/lifecycle-experimental-orange.svg" alt="Lifecycle: experimental"></a></li>
<li><a href="https://CRAN.R-project.org/package=prego" class="external-link"><img src="https://www.r-pkg.org/badges/version/prego" alt="CRAN status"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aviezer Lifshitz, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
