<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute quantile of pwm for a given interval size — gpwm_quantiles • prego</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute quantile of pwm for a given interval size — gpwm_quantiles"><meta name="description" content="Computes the quantile of the pwm for a given interval size by sampling random intervals from the genome, or using given intervals. The number of sequences to sample can be specified with n_sequences."><meta property="og:description" content="Computes the quantile of the pwm for a given interval size by sampling random intervals from the genome, or using given intervals. The number of sequences to sample can be specified with n_sequences."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">prego</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.9</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Compute quantile of pwm for a given interval size</h1>

      <div class="d-none name"><code>gpwm_quantiles.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Computes the quantile of the pwm for a given interval size by sampling random intervals from the genome, or using given intervals. The number of sequences to sample can be specified with <code>n_sequences</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gpwm_quantiles</span><span class="op">(</span></span>
<span>  <span class="va">size</span>,</span>
<span>  <span class="va">quantiles</span>,</span>
<span>  <span class="va">pssm</span>,</span>
<span>  bg_intervals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spat_min <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  spat_max <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  bidirect <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  prior <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  n_sequences <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  dist_from_edge <span class="op">=</span> <span class="fl">3000000</span>,</span>
<span>  chromosomes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  func <span class="op">=</span> <span class="st">"logSumExp"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-size">size<a class="anchor" aria-label="anchor" href="#arg-size"></a></dt>
<dd><p>size of the intervals to sample</p></dd>


<dt id="arg-quantiles">quantiles<a class="anchor" aria-label="anchor" href="#arg-quantiles"></a></dt>
<dd><p>quantiles to compute. See <code>quantile</code> for more details.</p></dd>


<dt id="arg-pssm">pssm<a class="anchor" aria-label="anchor" href="#arg-pssm"></a></dt>
<dd><p>PSSM matrix or data frame</p></dd>


<dt id="arg-bg-intervals">bg_intervals<a class="anchor" aria-label="anchor" href="#arg-bg-intervals"></a></dt>
<dd><p>(optional) an intervals set for the background. If not provided, random intervals will be used</p></dd>


<dt id="arg-spat">spat<a class="anchor" aria-label="anchor" href="#arg-spat"></a></dt>
<dd><p>a data frame with the spatial model (as returned from the <code>$spat</code> slot from the regression). Should contain a column called 'bin' and a column called 'spat_factor'.</p></dd>


<dt id="arg-spat-min">spat_min<a class="anchor" aria-label="anchor" href="#arg-spat-min"></a></dt>
<dd><p>the minimum position to use from the sequences. The default is 1.</p></dd>


<dt id="arg-spat-max">spat_max<a class="anchor" aria-label="anchor" href="#arg-spat-max"></a></dt>
<dd><p>the maximum position to use from the sequences. The default is the length of the sequences.</p></dd>


<dt id="arg-bidirect">bidirect<a class="anchor" aria-label="anchor" href="#arg-bidirect"></a></dt>
<dd><p>is the motif bi-directional. If TRUE, the reverse-complement of the motif will be used as well.</p></dd>


<dt id="arg-prior">prior<a class="anchor" aria-label="anchor" href="#arg-prior"></a></dt>
<dd><p>a prior probability for each nucleotide.</p></dd>


<dt id="arg-n-sequences">n_sequences<a class="anchor" aria-label="anchor" href="#arg-n-sequences"></a></dt>
<dd><p>number of sequences to sample in order to compute the quantiles. The default is 1e4.</p></dd>


<dt id="arg-dist-from-edge">dist_from_edge<a class="anchor" aria-label="anchor" href="#arg-dist-from-edge"></a></dt>
<dd><p>The minimum distance from the edge of the chromosome for a region to start or end(default: 3e6)</p></dd>


<dt id="arg-chromosomes">chromosomes<a class="anchor" aria-label="anchor" href="#arg-chromosomes"></a></dt>
<dd><p>The chromosomes to sample from (default: all chromosomes)</p></dd>


<dt id="arg-func">func<a class="anchor" aria-label="anchor" href="#arg-func"></a></dt>
<dd><p>the function to use to combine the PWMs for each sequence. Either 'logSumExp' or 'max'. The default is 'logSumExp'.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a named vector with the quantiles of the pwm for the given interval size.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tanaylab.github.io/misha/" class="external-link">misha</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">gdb.init_examples</span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pssm</span> <span class="op">&lt;-</span> <span class="va">JASPAR_motifs</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">motif</span> <span class="op">==</span> <span class="st">"JASPAR.CDX1"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">motif</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">gpwm_quantiles</span><span class="op">(</span><span class="fl">1000</span>, <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">0.1</span><span class="op">)</span>, <span class="va">pssm</span>, dist_from_edge <span class="op">=</span> <span class="fl">100</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aviezer Lifshitz, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

