<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ prego</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">prego</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="motif-regression">Motif Regression<a class="anchor" aria-label="anchor" href="#motif-regression"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="regress_pwm.html">regress_multiple_motifs()</a></code> <code><a href="regress_pwm.html">regress_pwm()</a></code>

        </dt>
        <dd>Perform a PWM regression</dd>
      </dl><dl><dt>

          <code><a href="regress_pwm.clusters.html">regress_pwm.clusters()</a></code>

        </dt>
        <dd>Run PWM regression on clusters</dd>
      </dl><dl><dt>

          <code><a href="regress_pwm.cv.html">regress_pwm.cv()</a></code>

        </dt>
        <dd>Cross-validate a PWM regression model</dd>
      </dl><dl><dt>

          <code><a href="regress_pwm.sample.html">regress_pwm.sample()</a></code>

        </dt>
        <dd>Run PWM regression on a sample of the data</dd>
      </dl><dl><dt>

          <code><a href="plot_regression_prediction.html">plot_regression_prediction()</a></code>

        </dt>
        <dd>Plot response variable averages vs the regression model's prediction</dd>
      </dl><dl><dt>

          <code><a href="plot_regression_prediction_binary.html">plot_regression_prediction_binary()</a></code>

        </dt>
        <dd>Plot the cumulative of the regression model's prediction stratified by the response variable</dd>
      </dl><dl><dt>

          <code><a href="plot_regression_qc.html">plot_regression_qc()</a></code>

        </dt>
        <dd>Plot the regression results</dd>
      </dl><dl><dt>

          <code><a href="plot_regression_qc_multi.html">plot_regression_qc_multi()</a></code>

        </dt>
        <dd>Plot the regression results for multiple motifs</dd>
      </dl><dl><dt>

          <code><a href="plot_spat_model.html">plot_spat_model()</a></code>

        </dt>
        <dd>Plot spatial model of the regression result</dd>
      </dl><dl><dt>

          <code><a href="export_multi_regression.html">export_multi_regression()</a></code>

        </dt>
        <dd>Export a multiple motif regression model</dd>
      </dl><dl><dt>

          <code><a href="export_regression_model.html">export_regression_model()</a></code>

        </dt>
        <dd>Export a motif regression model</dd>
      </dl><dl><dt>

          <code><a href="load_multi_regression.html">load_multi_regression()</a></code>

        </dt>
        <dd>Load a multiple motif regression model from a file</dd>
      </dl><dl><dt>

          <code><a href="load_regression_model.html">load_regression_model()</a></code>

        </dt>
        <dd>Load a motif regression model from a file</dd>
      </dl></div><div class="section level2">
      <h2 id="kmers">kmers<a class="anchor" aria-label="anchor" href="#kmers"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="screen_kmers.html">screen_kmers()</a></code>

        </dt>
        <dd>Screen for kmers</dd>
      </dl><dl><dt>

          <code><a href="generate_kmers.html">generate_kmers()</a></code>

        </dt>
        <dd>Generate kmers</dd>
      </dl><dl><dt>

          <code><a href="kmer_matrix.html">kmer_matrix()</a></code>

        </dt>
        <dd>Generate a kmer Matrix</dd>
      </dl><dl><dt>

          <code><a href="kmers_to_pssm.html">kmers_to_pssm()</a></code>

        </dt>
        <dd>Transform k-mers to PSSM (Position-Specific Scoring Matrix)</dd>
      </dl></div><div class="section level2">
      <h2 id="pssm-pwm-utils">PSSM/PWM utils<a class="anchor" aria-label="anchor" href="#pssm-pwm-utils"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="compute_pwm.html">compute_pwm()</a></code>

        </dt>
        <dd>Compute PWMs for a set of sequences given a PSSM matrix</dd>
      </dl><dl><dt>

          <code><a href="extract_pwm.html">extract_pwm_old()</a></code> <code><a href="extract_pwm.html">extract_pwm()</a></code>

        </dt>
        <dd>Extract pwm of sequences from a motif database</dd>
      </dl><dl><dt>

          <code><a href="gextract_pwm.html">gextract_pwm_old()</a></code> <code><a href="gextract_pwm.html">gextract_pwm()</a></code>

        </dt>
        <dd>Extract pwm of intervals from a motif database</dd>
      </dl><dl><dt>

          <code><a href="consensus_from_pssm.html">consensus_from_pssm()</a></code>

        </dt>
        <dd>Convert PSSM to consensus sequence</dd>
      </dl><dl><dt>

          <code><a href="pssm_diff.html">pssm_diff()</a></code>

        </dt>
        <dd>Compute KL divergence between two PSSMs</dd>
      </dl><dl><dt>

          <code><a href="pssm_match.html">pssm_match()</a></code>

        </dt>
        <dd>Match PSSM to a directory of motifs</dd>
      </dl><dl><dt>

          <code><a href="pssm_cor.html">pssm_cor()</a></code>

        </dt>
        <dd>Compute the correlation between two given PSSMs</dd>
      </dl><dl><dt>

          <code><a href="plot_pssm_logo.html">plot_pssm_logo()</a></code>

        </dt>
        <dd>Plot LOGO of the pssm result from the regression</dd>
      </dl><dl><dt>

          <code><a href="plot_pssm_logo_dataset.html">plot_pssm_logo_dataset()</a></code>

        </dt>
        <dd>Plot LOGO of pssm from dataset (e.g. "HOMER" or "JASPAR")</dd>
      </dl><dl><dt>

          <code><a href="all_motif_datasets.html">all_motif_datasets()</a></code>

        </dt>
        <dd>Get a data frame of all the motif datasets bundled with prego</dd>
      </dl><dl><dt>

          <code><a href="screen_pwm.html">screen_pwm()</a></code>

        </dt>
        <dd>Screen for motifs in a database given a response variable</dd>
      </dl><dl><dt>

          <code><a href="screen_pwm.clusters.html">screen_pwm.clusters()</a></code>

        </dt>
        <dd>Screen for motifs in a database for every cluster</dd>
      </dl><dl><dt>

          <code><a href="get_motif_pssm.html">get_motif_pssm()</a></code>

        </dt>
        <dd>Extract pssm of sequences from a motif database</dd>
      </dl><dl><dt>

          <code><a href="HOCOMOCO_motifs.html">HOCOMOCO_motifs</a></code>

        </dt>
        <dd>PSSMs from the HOCOMOCO motif database</dd>
      </dl><dl><dt>

          <code><a href="HOMER_motifs.html">HOMER_motifs</a></code>

        </dt>
        <dd>PSSMs from the HOMER motif database</dd>
      </dl><dl><dt>

          <code><a href="JASPAR_motifs.html">JASPAR_motifs</a></code>

        </dt>
        <dd>PSSMs from the JASPAR motif database</dd>
      </dl><dl><dt>

          <code><a href="JOLMA_motifs.html">JOLMA_motifs</a></code>

        </dt>
        <dd>PSSMs from the Jolma et al. motif database</dd>
      </dl><dl><dt>

          <code><a href="regress_pwm.html">regress_multiple_motifs()</a></code> <code><a href="regress_pwm.html">regress_pwm()</a></code>

        </dt>
        <dd>Perform a PWM regression</dd>
      </dl><dl><dt>

          <code><a href="compute_local_pwm.html">compute_local_pwm()</a></code>

        </dt>
        <dd>Compute local PWMs for a set of sequences given a PSSM matrix</dd>
      </dl><dl><dt>

          <code><a href="gextract_pwm.quantile.html">gextract_pwm.quantile()</a></code>

        </dt>
        <dd>Extract quantiles of pwm of intervals from a motif database</dd>
      </dl><dl><dt>

          <code><a href="gpwm_quantiles.html">gpwm_quantiles()</a></code>

        </dt>
        <dd>Compute quantile of pwm for a given interval size</dd>
      </dl><dl><dt>

          <code><a href="mask_sequences_by_pwm.html">mask_sequences_by_pwm()</a></code>

        </dt>
        <dd>Mask sequences by thresholding the PWM</dd>
      </dl><dl><dt>

          <code><a href="trim_pssm.html">trim_pssm()</a></code>

        </dt>
        <dd>Trim PSSM</dd>
      </dl><dl><dt>

          <code><a href="MOTIF_DB.html">MOTIF_DB</a></code>

        </dt>
        <dd>A MotifDB object with motifs from all bundled databases</dd>
      </dl><dl><dt>

          <code><a href="MotifDB-class.html">MotifDB-class</a></code>

        </dt>
        <dd>MotifDB Class</dd>
      </dl><dl><dt>

          <code><a href="as.data.frame-MotifDB-method.html">as.data.frame(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Convert a MotifDB object to a data frame</dd>
      </dl><dl><dt>

          <code><a href="as.matrix-MotifDB-method.html">as.matrix(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Convert a MotifDB object to a matrix</dd>
      </dl><dl><dt>

          <code><a href="initialize-MotifDB-method.html">initialize(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Initialize method for MotifDB objects</dd>
      </dl><dl><dt>

          <code><a href="length-MotifDB-method.html">length(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Get the length of a MotifDB object</dd>
      </dl><dl><dt>

          <code><a href="names-MotifDB-method.html">names(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Get the names of motifs in a MotifDB object</dd>
      </dl><dl><dt>

          <code><a href="plot-MotifDB-method.html">plot(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Plot a motif from a MotifDB object</dd>
      </dl><dl><dt>

          <code><a href="prior-MotifDB-method.html">prior(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Get the prior value from a MotifDB object</dd>
      </dl><dl><dt>

          <code><a href="prior-set-MotifDB-method.html">`prior&lt;-`(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Set a new prior for a MotifDB object</dd>
      </dl><dl><dt>

          <code><a href="show-MotifDB-method.html">show(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Show method for MotifDB objects</dd>
      </dl><dl><dt>

          <code><a href="sub-MotifDB-method.html">`[`(<i>&lt;MotifDB&gt;</i>)</a></code>

        </dt>
        <dd>Get specific motifs from the MotifDB</dd>
      </dl><dl><dt>

          <code><a href="motif_db_to_dataframe.html">motif_db_to_dataframe()</a></code>

        </dt>
        <dd>Convert a MotifDB object back to a tidy data frame</dd>
      </dl><dl><dt>

          <code><a href="prior-set.html">`prior&lt;-`()</a></code>

        </dt>
        <dd>Set the prior for a MotifDB object</dd>
      </dl><dl><dt>

          <code><a href="prior.html">prior()</a></code>

        </dt>
        <dd>Get or set the prior for a MotifDB object</dd>
      </dl><dl><dt>

          <code><a href="create_motif_db.html">create_motif_db()</a></code>

        </dt>
        <dd>Create a MotifDB object from a tidy data frame</dd>
      </dl><dl><dt>

          <code><a href="bits_per_pos.html">bits_per_pos()</a></code>

        </dt>
        <dd>Calculate the number of bits per position in a Position-Specific Scoring Matrix (PSSM).</dd>
      </dl><dl><dt>

          <code><a href="calc_seq_pwm.html">calc_seq_pwm()</a></code>

        </dt>
        <dd>Calculate Position Weight Matrix (PWM) Scores for DNA Sequences</dd>
      </dl><dl><dt>

          <code><a href="pssm_dataset_cor.html">pssm_dataset_cor()</a></code>

        </dt>
        <dd>Compute a correlation matrix for a pssm dataset</dd>
      </dl><dl><dt>

          <code><a href="pssm_rc.html">pssm_rc()</a></code>

        </dt>
        <dd>Reverse complement a PSSM</dd>
      </dl><dl><dt>

          <code><a href="pssm_to_kmer.html">pssm_to_kmer()</a></code>

        </dt>
        <dd>Transform PSSM (Position-Specific Scoring Matrix) to a KMER</dd>
      </dl><dl><dt>

          <code><a href="rc.html">rc()</a></code>

        </dt>
        <dd>Reverse Complement DNA Sequences</dd>
      </dl></div><div class="section level2">
      <h2 id="sequence-utils">Sequence utils<a class="anchor" aria-label="anchor" href="#sequence-utils"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="intervals_to_seq.html">intervals_to_seq()</a></code>

        </dt>
        <dd>Convert intervals to sequences</dd>
      </dl><dl><dt>

          <code><a href="calc_sequences_dinuc_dist.html">calc_sequences_dinuc_dist()</a></code>

        </dt>
        <dd>Calculate Dinucleotide Distribution in Sequences</dd>
      </dl><dl><dt>

          <code><a href="calc_sequences_dinucs.html">calc_sequences_dinucs()</a></code>

        </dt>
        <dd>Calculate Dinucleotide Counts for Sequences</dd>
      </dl><dl><dt>

          <code><a href="calc_sequences_trinuc_dist.html">calc_sequences_trinuc_dist()</a></code>

        </dt>
        <dd>Calculate Trinucleotide Distribution in Sequences</dd>
      </dl></div><div class="section level2">
      <h2 id="local-pwm">Local PWM<a class="anchor" aria-label="anchor" href="#local-pwm"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="gintervals.center_by_pssm.html">gintervals.center_by_pssm()</a></code>

        </dt>
        <dd>Center intervals by PSSM</dd>
      </dl><dl><dt>

          <code><a href="gextract.local_pwm.html">gextract.local_pwm()</a></code>

        </dt>
        <dd>Extracts local position weight matrix (PWM) scores for given intervals and a PWM.</dd>
      </dl><dl><dt>

          <code><a href="gextract.local_pwm_freq.html">gextract.local_pwm_freq()</a></code>

        </dt>
        <dd>Calculate the frequency of a position weight matrix (PWM) in a given set of intervals</dd>
      </dl></div><div class="section level2">
      <h2 id="other">other<a class="anchor" aria-label="anchor" href="#other"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="motif_enrichment.html">motif_enrichment()</a></code>

        </dt>
        <dd>Calculate motif enrichment</dd>
      </dl><dl><dt>

          <code><a href="sample_quantile_matched_rows.html">sample_quantile_matched_rows()</a></code>

        </dt>
        <dd>Sample rows respecting quantiles of a reference distribution</dd>
      </dl><dl><dt>

          <code><a href="set_parallel.html">set_parallel()</a></code>

        </dt>
        <dd>Set parallel threads</dd>
      </dl><dl><dt>

          <code><a href="cluster_mat_example.html">cluster_mat_example</a></code> <code><a href="cluster_mat_example.html">cluster_sequences_example</a></code> <code><a href="cluster_mat_example.html">clusters_example</a></code> <code><a href="cluster_mat_example.html">response_mat_example</a></code> <code><a href="cluster_mat_example.html">sequences_example</a></code>

        </dt>
        <dd>Dataset examples</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aviezer Lifshitz, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

