<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Run PWM regression on clusters"><title>Run PWM regression on clusters — regress_pwm.clusters • prego</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run PWM regression on clusters — regress_pwm.clusters"><meta property="og:description" content="Run PWM regression on clusters"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">prego</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Run PWM regression on clusters</h1>
      
      <div class="d-none name"><code>regress_pwm.clusters.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Run PWM regression on clusters</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">regress_pwm.clusters</span><span class="op">(</span></span>
<span>  <span class="va">sequences</span>,</span>
<span>  <span class="va">clusters</span>,</span>
<span>  use_sample <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  match_with_db <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sample_frac <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  final_metric <span class="op">=</span> <span class="st">"ks"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"prego.parallel"</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>sequences</dt>
<dd><p>A vector of DNA sequences ('A', 'T', 'C' or 'G'. Will go through <code>toupper</code>)</p></dd>


<dt>clusters</dt>
<dd><p>a vector with the cluster assignments for each sequence</p></dd>


<dt>use_sample</dt>
<dd><p>whether to use sampled optimization or not (default: FALSE).</p></dd>


<dt>match_with_db</dt>
<dd><p>match the resulting PWMs with motif databases using <code>pssm_match</code>.
This would add a column named 'db_match' to the stats data frame, together with 'pred_mat_db' with the
database motif predictions, and and 'db_dataset' which is similiar to 'motif_dataset' for the database motifs.
Note that the closest match is returned, even if it is not similar enough in absolute terms.
Also, the match is done between the rsulting regression <em>pssm</em> and the pssms in the databse - in order to find the best motif in the database which explain the clusters, use <code>screen_pwm.clusters</code>.</p></dd>


<dt>sample_frac</dt>
<dd><p>a vector of two numbers, specifying the fraction of
sequences to use in when sampling for the sequences which are not
in the cluster (first number) and in the cluster (second number). If NULL -</p></dd>


<dt>sample_ratio</dt>
<dd><p>When <code>sample_frac</code> is NULL, the number of sequences not in the cluster would be equal to <code>sample_ratio</code> times the number of sequences in the cluster.</p></dd>


<dt>final_metric</dt>
<dd><p>metric to use in order to choose the best motif. One of 'ks' or 'r2'. Note that unlike <code>score_metric</code> which is used in the regression itself, this metric is used only for choosing the best motif out of all the runs on the sampled dataset.</p></dd>


<dt>parallel</dt>
<dd><p>whether to run optimization in parallel. use <code>set_parallel</code>
to set the number of cores to use.</p></dd>


<dt>...</dt>
<dd><p>Arguments passed on to <code><a href="regress_pwm.html">regress_pwm</a></code>, <code><a href="regress_pwm.sample.html">regress_pwm.sample</a></code></p><dl><dt><code>response</code></dt>
<dd><p>A matrix of response variables - number of rows should equal the number of sequences</p></dd>

    <dt><code>motif</code></dt>
<dd><p>Initial motif to start the regression from. Can be either a string with a kmer where the character "*" indicates a
wildcard or a data frame with a pre-computed PSSM (see thre slot <code>pssm</code> in the return value of this function).
If NULL - a K-mer screen would be performed in order to find the best kmer for initialization.</p></dd>

    <dt><code>motif_length</code></dt>
<dd><p>Length of the seed motif. If the motif is shorter than this, it will be extended by wildcards (stars). Note that If the motif is longer than this, it will <em>not</em> be truncated.</p></dd>

    <dt><code>score_metric</code></dt>
<dd><p>metric to use for optimizing the PWM. One of "r2" or "ks". When using "ks" the response variable should be a single vector of 0 and 1.</p></dd>

    <dt><code>bidirect</code></dt>
<dd><p>is the motif bi-directional. If TRUE, the reverse-complement of the motif will be used as well.</p></dd>

    <dt><code>spat_min</code></dt>
<dd><p>start of the spatial model from the beginning of the sequence (in bp)</p></dd>

    <dt><code>spat_max</code></dt>
<dd><p>end of the spatial model from the beginning of the sequence (in bp). If NULL - the spatial model
would end at the end of the sequence.</p></dd>

    <dt><code>spat_bin</code></dt>
<dd><p>size of the spatial bin (in bp).</p></dd>

    <dt><code>spat_model</code></dt>
<dd><p>a previously computed spatial model (see <code>spat</code>) in the return value of this function. This can only be used when <code>motif</code> is a previously computed PSSM.</p></dd>

    <dt><code>improve_epsilon</code></dt>
<dd><p>minimum improve in the objective function to continue the optimization</p></dd>

    <dt><code>min_nuc_prob</code></dt>
<dd><p>minimum nucleotide probability in every iteration</p></dd>

    <dt><code>unif_prior</code></dt>
<dd><p>uniform prior for nucleotide probabilities</p></dd>

    <dt><code>include_response</code></dt>
<dd><p>include the response in the resulting list (default: TRUE)</p></dd>

    <dt><code>verbose</code></dt>
<dd><p>show verbose messages.</p></dd>

    <dt><code>seed</code></dt>
<dd><p>random seed</p></dd>

    <dt><code>consensus_single_thresh,consensus_double_thresh</code></dt>
<dd><p>thresholds for the consensus sequence calculation
(single and double nucleotides)</p></dd>

    <dt><code>motif_dataset</code></dt>
<dd><p>a data frame with PSSMs ('A', 'C', 'G' and 'T' columns), with an additional column 'motif' containing the motif name, for example <code>HOMER_motifs</code>, <code>JASPAR_motifs</code> or all_motif_datasets(). By default all_motif_datasets() would be used.</p></dd>

    <dt><code>multi_kmers</code></dt>
<dd><p>if TRUE, different candidates of kmers would be regressed in order to find the best seed according to <code>final_metric</code>.</p></dd>

    <dt><code>kmer_length</code></dt>
<dd><p>a vector of kmer lengths to screen in order to find the best seed motif.</p></dd>

    <dt><code>max_cands</code></dt>
<dd><p>maximum number of kmer candidates to try.</p></dd>

    <dt><code>motif_num</code></dt>
<dd><p>Number of motifs to infer. When <code>motif_num</code> &gt; 1, the function would run <code>motif_num</code> times, each time on the residuals of a linear model of all the previous runs (see <code>smooth_k</code> parameter). The best motif is then returned, while all the others are stored at 'models' in the return value.</p></dd>

    <dt><code>smooth_k</code></dt>
<dd><p>k for smoothing the predictions of each model in order to compute the resiuals when <code>motif_num</code> &gt; 1. The residulas are computed as <code>response</code> - running mean of size 'k' of the current model.</p></dd>

    <dt><code>is_train</code></dt>
<dd><p>a boolean vector that determine which subset of sequences to use when screening</p></dd>

    <dt><code>min_gap,max_gap</code></dt>
<dd><p>the length of a gap to be considered in the pattern. Only one gap, of length min_gap:max_gap, is being used, and is located anywhere in the motif. Note that this greatly expand the search space (and increase multiple tesing severly).</p></dd>

    <dt><code>sample_idxs</code></dt>
<dd><p>indices of the sequences to use. If NULL, the indices would be sampled using <code>sample_frac</code>.</p></dd>

  
</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a list with the following elements:</p><ul><li><p>models: a list with the models for each cluster</p></li>
<li><p>cluster_mat: an indicator matrix with the cluster assignments</p></li>
<li><p>pred_mat: a matrix with the predicted pwm for each sequence (rows) and cluster (columns)</p></li>
<li><p>motif_dataset: a data frame with the PSSMs for each cluster</p></li>
<li><p>spat_dataset: a data frame with the spatial model for each cluster</p></li>
<li><p>stats: a data frame with statistics for each cluster</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">regress_pwm.clusters</span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">clusters_example</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running regression for <span style="color: #0000BB;">5</span> clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Matching with motif databases</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pred_mat</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                           c100      c111       c29        c5        c6</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr1_5108540_5108840 -17.28819 -15.88604 -16.43380 -16.61165 -16.84203</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr1_7313460_7313760 -14.69661 -17.02461 -16.51139 -16.67072 -17.51148</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr1_7778180_7778480 -15.75232 -15.11354 -16.80705 -16.65310 -16.69591</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr1_9381040_9381340 -16.03874 -15.04809 -15.20886 -16.47437 -16.01101</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr1_9921360_9921660 -14.85159 -17.90363 -15.35405 -16.13744 -16.07321</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> chr1_9931940_9932240 -16.32993 -17.16320 -14.44998 -16.72751 -16.53625</span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">stats</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # A tibble: 5 x 7</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   cluster consensus      ks_D        r2      seed_motif              db_match</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    c100   AT***TC 0.7024075 0.3852463 ***ATCCATCA**** JASPAR.MEIS1.MA1639.1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    c111   Y*RTAAA 0.8479663 0.4978355 ***CAATTAAC****   JOLMA.CDX1_mono_DBD</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     c29 T*A***W*T 0.8507039 0.5581745 ***TAATCATT****            HOMER.Hnf1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      c5      GATA 0.5887910 0.2227806 ***CTGATAAG****      JASPAR.GATA1/2/3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      c6      TATC 0.6160721 0.2783910 ***TCTTATCT****          JASPAR.GATAd</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   db_match_dist</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1      0.244724</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2      0.152716</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3      0.340463</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      0.198050</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      0.219126</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">models</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>`guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")` instead.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>`guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")` instead.</span>
<span class="r-plt img"><img src="regress_pwm.clusters-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># multiple motifs per cluster</span></span></span>
<span class="r-in"><span><span class="va">res_multi</span> <span class="op">&lt;-</span> <span class="fu">regress_pwm.clusters</span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">clusters_example</span>, motif_num <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running regression for <span style="color: #0000BB;">5</span> clusters</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in do.ply(i):</span> task 1 failed - "Multiple motifs are not supported yet"</span>
<span class="r-in"><span><span class="va">res_multi</span><span class="op">$</span><span class="va">multi_stats</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in eval(expr, envir, enclos):</span> object 'res_multi' not found</span>
<span class="r-in"><span><span class="fu"><a href="plot_regression_qc_multi.html">plot_regression_qc_multi</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">models</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in plot_regression_qc_multi(res$models[[1]], title = names(res$models)[1]):</span> could not find function "plot_regression_qc_multi"</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Aviezer Lifshitz, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer></div>

  

  

  </body></html>

