<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Run PWM regression on clusters — regress_pwm.clusters • prego</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run PWM regression on clusters — regress_pwm.clusters"><meta name="description" content="Run PWM regression on clusters"><meta property="og:description" content="Run PWM regression on clusters"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">prego</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Run PWM regression on clusters</h1>

      <div class="d-none name"><code>regress_pwm.clusters.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Run PWM regression on clusters</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">regress_pwm.clusters</span><span class="op">(</span></span>
<span>  <span class="va">sequences</span>,</span>
<span>  <span class="va">clusters</span>,</span>
<span>  use_sample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  match_with_db <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  screen_db <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sample_frac <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  final_metric <span class="op">=</span> <span class="st">"ks"</span>,</span>
<span>  parallel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"prego.parallel"</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  use_sge <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dataset <span class="op">=</span> <span class="fu"><a href="all_motif_datasets.html">all_motif_datasets</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  motifs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  min_D <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  prior <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  alternative <span class="op">=</span> <span class="st">"two.sided"</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sequences">sequences<a class="anchor" aria-label="anchor" href="#arg-sequences"></a></dt>
<dd><p>A vector of DNA sequences ('A', 'T', 'C' or 'G'. Will go through <code>toupper</code>). Please make sure that the sequences are long enough to cover <code>spat_num_bins</code> * <code>spat_bin_size</code> bp, and that they are centered around the motif/signal.</p></dd>


<dt id="arg-clusters">clusters<a class="anchor" aria-label="anchor" href="#arg-clusters"></a></dt>
<dd><p>a vector with the cluster assignments for each sequence</p></dd>


<dt id="arg-use-sample">use_sample<a class="anchor" aria-label="anchor" href="#arg-use-sample"></a></dt>
<dd><p>whether to use sampled optimization or not (default: FALSE).</p></dd>


<dt id="arg-match-with-db">match_with_db<a class="anchor" aria-label="anchor" href="#arg-match-with-db"></a></dt>
<dd><p>match the resulting PWMs with motif databases using <code>pssm_match</code>.
This would add a column named 'db_match' to the stats data frame, together with 'pred_mat_db' with the
database motif predictions, and 'db_dataset' which is similar to 'motif_dataset' for the database motifs.
Note that the closest match is returned, even if it is not similar enough in absolute terms.
Also, the match is done between the resulting regression <em>pssm</em> and the pssms in the database - in order to find the best motif in the database set <code>screen_db=TRUE</code>.</p></dd>


<dt id="arg-screen-db">screen_db<a class="anchor" aria-label="anchor" href="#arg-screen-db"></a></dt>
<dd><p>screen for the best motif in the database which explains the clusters. See <code>screen_pwm.clusters</code>.</p></dd>


<dt id="arg-sample-frac">sample_frac<a class="anchor" aria-label="anchor" href="#arg-sample-frac"></a></dt>
<dd><p>a vector of two numbers, specifying the fraction of
sequences to use in when sampling for the sequences which are not
in the cluster (first number) and in the cluster (second number). If NULL -</p></dd>


<dt id="arg-sample-ratio">sample_ratio<a class="anchor" aria-label="anchor" href="#arg-sample-ratio"></a></dt>
<dd><p>When <code>sample_frac</code> is NULL, the number of sequences not in the cluster would be equal to <code>sample_ratio</code> times the number of sequences in the cluster.</p></dd>


<dt id="arg-final-metric">final_metric<a class="anchor" aria-label="anchor" href="#arg-final-metric"></a></dt>
<dd><p>metric to use in order to choose the best motif. One of 'ks' or 'r2'. Note that unlike <code>score_metric</code> which is used in the regression itself, this metric is used only for choosing the best motif out of all the runs on the sampled dataset. If NULL - 'ks' would be used for binary response and 'r2' for continuous response.</p></dd>


<dt id="arg-parallel">parallel<a class="anchor" aria-label="anchor" href="#arg-parallel"></a></dt>
<dd><p>whether to run optimization in parallel. use <code>set_parallel</code>
to set the number of cores to use.</p></dd>


<dt id="arg-use-sge">use_sge<a class="anchor" aria-label="anchor" href="#arg-use-sge"></a></dt>
<dd><p>use the function <code>gcluster.run2</code> from the misha.ext package to run the optimization on a SGE cluster. Only relevant if the <code>misha.ext</code> package is installed. Note that <code>gcluster.run2</code> writes the current
environment before starting the parallelization, so it is better to run this function in a clean environment.
Also, Note that 'prego' needs to be installed in order for this to work, i.e. you cannot use <code><a href="https://devtools.r-lib.org/reference/load_all.html" class="external-link">devtools::load_all()</a></code> or <code><a href="https://pkgload.r-lib.org/reference/load_all.html" class="external-link">pkgload::load_all()</a></code> to load the package.</p></dd>


<dt id="arg-dataset">dataset<a class="anchor" aria-label="anchor" href="#arg-dataset"></a></dt>
<dd><p>a data frame with PSSMs ('A', 'C', 'G' and 'T' columns), with an additional column 'motif' containing the motif name, for example <code>HOMER_motifs</code> or <code>JASPAR_motifs</code>, or <code><a href="all_motif_datasets.html">all_motif_datasets()</a></code>, or a MotifDB object.</p></dd>


<dt id="arg-motifs">motifs<a class="anchor" aria-label="anchor" href="#arg-motifs"></a></dt>
<dd><p>names of specific motifs to extract from the dataset</p></dd>


<dt id="arg-min-d">min_D<a class="anchor" aria-label="anchor" href="#arg-min-d"></a></dt>
<dd><p>minimum distance to consider a match</p></dd>


<dt id="arg-prior">prior<a class="anchor" aria-label="anchor" href="#arg-prior"></a></dt>
<dd><p>a prior probability for each nucleotide.</p></dd>


<dt id="arg-alternative">alternative<a class="anchor" aria-label="anchor" href="#arg-alternative"></a></dt>
<dd><p>alternative hypothesis for the KS test. Can be "two.sided", "less" or "greater"</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Arguments passed on to <code><a href="regress_pwm.html">regress_pwm</a></code>, <code><a href="regress_pwm.sample.html">regress_pwm.sample</a></code></p><dl><dt><code>response</code></dt>
<dd><p>A matrix of response variables - number of rows should equal the number of sequences</p></dd>

    <dt><code>motif</code></dt>
<dd><p>Initial motif to start the regression from. Can be either a string with a kmer where the character "*" indicates a
wildcard or a data frame with a pre-computed PSSM (see the slot <code>pssm</code> in the return value of this function).
If NULL - a K-mer screen would be performed in order to find the best kmer for initialization. If <code>init_from_dataset</code> is TRUE, the regression would be initialized from the PSSM of the best motif in the dataset.</p></dd>

    <dt><code>init_from_dataset</code></dt>
<dd><p>initialize the regression from the PSSM of the best motif in <code>motif_dataset</code>, using <code>final_metric</code> as the metric. If TRUE, the <code>motif</code> parameter would be ignored. See <code><a href="screen_pwm.html">screen_pwm</a></code> for more details.</p></dd>

    <dt><code>motif_length</code></dt>
<dd><p>Length of the seed motif. If the motif is shorter than this, it will be extended by wildcards (stars). Note that If the motif is longer than this, it will <em>not</em> be truncated.</p></dd>

    <dt><code>score_metric</code></dt>
<dd><p>metric to use for optimizing the PWM. One of "r2" or "ks". When using "ks" the response variable should be a single vector of 0 and 1.</p></dd>

    <dt><code>bidirect</code></dt>
<dd><p>is the motif bi-directional. If TRUE, the reverse-complement of the motif will be used as well.</p></dd>

    <dt><code>spat_bin_size</code></dt>
<dd><p>size of the spatial bin (in bp).</p></dd>

    <dt><code>spat_num_bins</code></dt>
<dd><p>number of spatial bins. Please make sure that the sequences are long enough to cover this number of bins. bp outside of spat_bin_size * spat_num_bins would be ignored. If <code>bidirect</code> is TRUE, the number of bins should be odd as 'prego' symmetrizes the motif around the center bin.</p></dd>

    <dt><code>spat_model</code></dt>
<dd><p>a previously computed spatial model (see <code>spat</code>) in the return value of this function.</p></dd>

    <dt><code>improve_epsilon</code></dt>
<dd><p>minimum improve in the objective function to continue the optimization</p></dd>

    <dt><code>min_nuc_prob</code></dt>
<dd><p>minimum nucleotide probability in every iteration</p></dd>

    <dt><code>unif_prior</code></dt>
<dd><p>uniform prior for nucleotide probabilities</p></dd>

    <dt><code>include_response</code></dt>
<dd><p>include the response in the resulting list (default: TRUE)</p></dd>

    <dt><code>verbose</code></dt>
<dd><p>show verbose messages.</p></dd>

    <dt><code>seed</code></dt>
<dd><p>random seed</p></dd>

    <dt><code>consensus_single_thresh,consensus_double_thresh</code></dt>
<dd><p>thresholds for the consensus sequence calculation
(single and double nucleotides)</p></dd>

    <dt><code>motif_dataset</code></dt>
<dd><p>a data frame with PSSMs ('A', 'C', 'G' and 'T' columns), with an additional column 'motif' containing the motif name, for example <code>HOMER_motifs</code>, <code>JASPAR_motifs</code> or all_motif_datasets(). By default all_motif_datasets() would be used.</p></dd>

    <dt><code>multi_kmers</code></dt>
<dd><p>if TRUE, different candidates of kmers would be regressed in order to find the best seed according to <code>final_metric</code>.</p></dd>

    <dt><code>kmer_length</code></dt>
<dd><p>a vector of kmer lengths to screen in order to find the best seed motif.</p></dd>

    <dt><code>max_cands</code></dt>
<dd><p>maximum number of kmer candidates to try.</p></dd>

    <dt><code>motif_num</code></dt>
<dd><p>Number of motifs to infer. When <code>motif_num</code> &gt; 1, the function would run <code>motif_num</code> times, each time on the residuals of a linear model of all the previous runs (see <code>smooth_k</code> parameter). The best motif is then returned, while all the others are stored at 'models' in the return value.</p></dd>

    <dt><code>smooth_k</code></dt>
<dd><p>k for smoothing the predictions of each model in order to compute the residuals when <code>motif_num</code> &gt; 1. The residuals are computed as <code>response</code> - running mean of size 'k' of the current model.</p></dd>

    <dt><code>min_kmer_cor</code></dt>
<dd><p>minimal correlation between the kmer and the response in order to use it as a seed.</p></dd>

    <dt><code>internal_num_folds</code></dt>
<dd><p>number of folds to use in the internal cross-validation.</p></dd>

    <dt><code>sample_for_kmers</code></dt>
<dd><p>Use a random sample of the dataset in order to find the best kmer. This is useful when the dataset is very large and the kmer screen would take a long time. Note that the final regression would be performed on the entire dataset. Only relevant when <code>multi_kmers</code> is TRUE.</p></dd>

    <dt><code>sample_idxs</code></dt>
<dd><p>indices of the sequences to use for the kmer screen. If NULL, a random sample would be used.</p></dd>

    <dt><code>log_energy</code></dt>
<dd><p>transform the energy to log scale on each iteration.</p></dd>

    <dt><code>energy_func</code></dt>
<dd><p>a function to transform the energy at each iteration. Should accept a numeric vector and return a numeric vector. e.g. <code>log</code> or <code>function(x) x^2</code>. Note that the range of the input energies is between 0 and 1 (the probability of the motif in the sequence), so if you inferred the function using the the returned energies (which are in log scale) you should make sure that the function first log transforms using <code>log_energy=TRUE</code>.</p></dd>

    <dt><code>xmin,xmax,npts</code></dt>
<dd><p>range for the energy function and the number of points to use for its interpolation.</p></dd>

    <dt><code>energy_func_generator</code></dt>
<dd><p>a function to generate the energy function when regressing multiple motifs. Should accept the result of the previous iteration + the original response and return a function similar to <code>energy_func</code>. e.g. <code>
function(prev_reg, resp) {
       df &lt;- data.frame(x = prev_reg$pred, y = resp)
       fn_gam &lt;- as.formula("y ~ s(x, k=3, bs='cr')")
       model &lt;- mgcv::gam(fn_gam, family = binomial(link = "logit"), data = df, method="REML")
       function(z){
           mgcv::predict.gam(object = model, newdata = data.frame(x = z))
}}</code>.
When this parameter is not NULL, energy_func_generator would create an energy function and then run another step of regression initialized with the previous motif with <code>energy_func</code> as the energy function. This is useful when the energy function is not monotonic, for example - one might want to use a gam model to fit the energy function like in the example above.</p></dd>

    <dt><code>optimize_pwm</code></dt>
<dd><p>optimize the PWM model (Default: TRUE). If FALSE, the PWM model would be used as the initial model for the spatial model.</p></dd>

    <dt><code>optimize_spat</code></dt>
<dd><p>optimize the spatial model (Default: TRUE). If FALSE, the spatial model would be used as the initial model for the PWM model.</p></dd>

    <dt><code>kmer_sequence_length</code></dt>
<dd><p>the length of the sequence to use for the kmer screen. If NULL, the entire sequence would be used.</p></dd>

    <dt><code>symmetrize_spat</code></dt>
<dd><p>if TRUE, the spatial model would be symmetrized around the center bin. Default: TRUE.</p></dd>

    <dt><code>min_gap,max_gap</code></dt>
<dd><p>the length of a gap to be considered in the pattern. Only one gap, of length min_gap:max_gap, is being used, and is located anywhere in the motif. Note that this greatly expand the search space (and increase multiple testing severely).</p></dd>


</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list with the following elements:</p><dl><dt>models: </dt>
<dd><p>a list with the models for each cluster</p></dd>

<dt>cluster_mat: </dt>
<dd><p>an indicator matrix with the cluster assignments</p></dd>

<dt>pred_mat: </dt>
<dd><p>a matrix of the energies of the predicted motifs per cluster (columns) in each sequence (rows)</p></dd>

<dt>motif_dataset: </dt>
<dd><p>a data frame with the PSSMs for each cluster</p></dd>

<dt>spat_dataset: </dt>
<dd><p>a data frame with the spatial model for each cluster</p></dd>

<dt>stats: </dt>
<dd><p>a data frame with statistics for each cluster</p></dd>


</dl></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">regress_pwm.clusters</span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">clusters_example</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pred_mat</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">stats</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">models</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># multiple motifs per cluster</span></span></span>
<span class="r-in"><span><span class="va">res_multi</span> <span class="op">&lt;-</span> <span class="fu">regress_pwm.clusters</span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">clusters_example</span>, motif_num <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res_multi</span><span class="op">$</span><span class="va">multi_stats</span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_regression_qc_multi.html">plot_regression_qc_multi</a></span><span class="op">(</span><span class="va">res_multi</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res_multi</span><span class="op">$</span><span class="va">models</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># screen also for the best motif in the database</span></span></span>
<span class="r-in"><span><span class="va">res_screen</span> <span class="op">&lt;-</span> <span class="fu">regress_pwm.clusters</span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">clusters_example</span>, screen_db <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Using sampled optimization</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running regression for <span style="color: #0000BB;">5</span> clusters</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Matching with motif databases</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Screening motif databases for <span style="color: #0000BB;">5</span> clusters</span>
<span class="r-in"><span><span class="va">res_screen</span><span class="op">$</span><span class="va">stats</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> # A tibble: 5 x 9</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   cluster consensus      ks_D        r2      seed_motif</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    c100  WGA***AT 0.6530423 0.3224333 ***TGAT*GATG***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    c111    Y*RTAA 0.8413087 0.5127852 ***CAATTAAC****</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     c29 T*A***W*Y 0.8219863 0.5360583 ***TTAA*CATT***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4      c5      &lt;NA&gt; 0.5633821 0.1974345 ***AATCA*TAA***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5      c6       ATC 0.5971442 0.2654848 ***TA*CTTATC***</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                        db_match db_match_cor   ks_D_db</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 HOCOMOCO.PDX1_HUMAN.H11MO.0.A    0.7996138 0.6337245</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2           JOLMA.CDX2_mono_DBD    0.8912204 0.7803839</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          JASPAR.RFX1.MA0365.1    0.7338186 0.8606183</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4        JOLMA.NKX6-1_mono_full    0.6767482 0.5793942</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                  JASPAR.GATA2    0.7807456 0.5790557</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                         db_motif</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1                    HOMER.HOXA2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2                    JASPAR.CDX1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 HOCOMOCO.HNF1B_HUMAN.H11MO.0.A</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4                    HOMER.Gata2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5                   JASPAR.GATA5</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res_screen</span><span class="op">$</span><span class="va">models</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">res_screen</span><span class="op">$</span><span class="va">models</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>All aesthetics have length 1, but the data has 2359 rows.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Please consider using `annotate()` or provide this layer with data containing a</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span>   single row.</span>
<span class="r-plt img"><img src="regress_pwm.clusters-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Aviezer Lifshitz, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

