<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The optimization would be performed with a sampled dataset of size sample_frac, or explicit sampled indices sample_idxs. Note that multi_kmers is TRUE by default."><title>Run PWM regression on a sample of the data — regress_pwm.sample • prego</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Run PWM regression on a sample of the data — regress_pwm.sample"><meta property="og:description" content="The optimization would be performed with a sampled dataset of size sample_frac, or explicit sampled indices sample_idxs. Note that multi_kmers is TRUE by default."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">prego</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Run PWM regression on a sample of the data</h1>
      
      <div class="d-none name"><code>regress_pwm.sample.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The optimization would be performed with a sampled dataset of size <code>sample_frac</code>, or explicit sampled indices <code>sample_idxs</code>. Note that <code>multi_kmers</code> is TRUE by default.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">regress_pwm.sample</span><span class="op">(</span></span>
<span>  <span class="va">sequences</span>,</span>
<span>  <span class="va">response</span>,</span>
<span>  bidirect <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  spat_min <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  spat_max <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_response <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  motif_num <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  multi_kmers <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  sample_frac <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_idxs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  sample_ratio <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  parallel <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">getOption</a></span><span class="op">(</span><span class="st">"prego.parallel"</span>, <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>  match_with_db <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  screen_db <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  motif_dataset <span class="op">=</span> <span class="fu"><a href="all_motif_datasets.html">all_motif_datasets</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">60427</span>,</span>
<span>  final_metric <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  unif_prior <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>sequences</dt>
<dd><p>A vector of DNA sequences ('A', 'T', 'C' or 'G'. Will go through <code>toupper</code>)</p></dd>


<dt>response</dt>
<dd><p>A matrix of response variables - number of rows should equal the number of sequences</p></dd>


<dt>bidirect</dt>
<dd><p>is the motif bi-directional. If TRUE, the reverse-complement of the motif will be used as well.</p></dd>


<dt>spat_min</dt>
<dd><p>start of the spatial model from the beginning of the sequence (in bp)</p></dd>


<dt>spat_max</dt>
<dd><p>end of the spatial model from the beginning of the sequence (in bp). If NULL - the spatial model
would end at the end of the sequence.</p></dd>


<dt>include_response</dt>
<dd><p>include the response in the resulting list (default: TRUE)</p></dd>


<dt>motif_num</dt>
<dd><p>Number of motifs to infer. When <code>motif_num</code> &gt; 1, the function would run <code>motif_num</code> times, each time on the residuals of a linear model of all the previous runs (see <code>smooth_k</code> parameter). The best motif is then returned, while all the others are stored at 'models' in the return value.</p></dd>


<dt>multi_kmers</dt>
<dd><p>if TRUE, different candidates of kmers would be regressed in order to find the best seed according to <code>final_metric</code>.</p></dd>


<dt>sample_frac</dt>
<dd><p>fraction of the dataset to sample. When <code>response</code> is categorical (0 and 1), the sampling would be stratified by the category, i.e. <code>sample_frac</code> can be a vector of length 2 with the fraction of 0 and 1 responses to sample respectively.
If NULL - the default would be 0.1 for continuous variables, and for binary variables - the number of 0 responses would be equal to <code>sample_ratio</code> times the number of 1 responses.</p></dd>


<dt>sample_idxs</dt>
<dd><p>indices of the sequences to use. If NULL, the indices would be sampled using <code>sample_frac</code>.</p></dd>


<dt>sample_ratio</dt>
<dd><p>ratio between the '1' category and the '0' category in the sampled dataset. Relevant only when <code>sample_frac</code> is NULL.</p></dd>


<dt>parallel</dt>
<dd><p>whether to run optimization in parallel. use <code>set_parallel</code>
to set the number of cores to use.</p></dd>


<dt>match_with_db</dt>
<dd><p>match the resulting PWMs with motif databases using <code>pssm_match</code>. Note that the closest match
is returned, even if it is not similar enough in absolute terms.</p></dd>


<dt>screen_db</dt>
<dd><p>Screen <code>motif_dataset</code> using <code>screen_pwm</code> and use the best motif as the initial motif. If TRUE, the following fields would be added to the return value:
"db_motif", "db_motif_pred", "db_motif_pssm" and "db_motif_score".</p></dd>


<dt>motif_dataset</dt>
<dd><p>a data frame with PSSMs ('A', 'C', 'G' and 'T' columns), with an additional column 'motif' containing the motif name, for example <code>HOMER_motifs</code>, <code>JASPAR_motifs</code> or all_motif_datasets(). By default all_motif_datasets() would be used.</p></dd>


<dt>seed</dt>
<dd><p>random seed</p></dd>


<dt>final_metric</dt>
<dd><p>metric to use in order to choose the best motif. One of 'ks' or 'r2'. Note that unlike <code>score_metric</code> which is used in the regression itself, this metric is used only for choosing the best motif out of all the runs on the sampled dataset. If NULL - 'ks' would be used for binary response and 'r2' for continuous response.</p></dd>


<dt>unif_prior</dt>
<dd><p>uniform prior for nucleotide probabilities</p></dd>


<dt>...</dt>
<dd><p>Arguments passed on to <code><a href="regress_pwm.html">regress_pwm</a></code>, <code><a href="screen_kmers.html">screen_kmers</a></code></p><dl><dt><code>motif</code></dt>
<dd><p>Initial motif to start the regression from. Can be either a string with a kmer where the character "*" indicates a
wildcard or a data frame with a pre-computed PSSM (see the slot <code>pssm</code> in the return value of this function).
If NULL - a K-mer screen would be performed in order to find the best kmer for initialization. If <code>init_from_dataset</code> is TRUE, the regression would be initialized from the PSSM of the best motif in the dataset.</p></dd>

    <dt><code>init_from_dataset</code></dt>
<dd><p>initialize the regression from the PSSM of the best motif in <code>motif_dataset</code>, using <code>final_metric</code> as the metric. If TRUE, the <code>motif</code> parameter would be ignored. See <code><a href="screen_pwm.html">screen_pwm</a></code> for more details.</p></dd>

    <dt><code>motif_length</code></dt>
<dd><p>Length of the seed motif. If the motif is shorter than this, it will be extended by wildcards (stars). Note that If the motif is longer than this, it will <em>not</em> be truncated.</p></dd>

    <dt><code>score_metric</code></dt>
<dd><p>metric to use for optimizing the PWM. One of "r2" or "ks". When using "ks" the response variable should be a single vector of 0 and 1.</p></dd>

    <dt><code>spat_bin</code></dt>
<dd><p>size of the spatial bin (in bp).</p></dd>

    <dt><code>spat_model</code></dt>
<dd><p>a previously computed spatial model (see <code>spat</code>) in the return value of this function. This can only be used when <code>motif</code> is a previously computed PSSM.</p></dd>

    <dt><code>improve_epsilon</code></dt>
<dd><p>minimum improve in the objective function to continue the optimization</p></dd>

    <dt><code>min_nuc_prob</code></dt>
<dd><p>minimum nucleotide probability in every iteration</p></dd>

    <dt><code>verbose</code></dt>
<dd><p>show verbose messages.</p></dd>

    <dt><code>consensus_single_thresh,consensus_double_thresh</code></dt>
<dd><p>thresholds for the consensus sequence calculation
(single and double nucleotides)</p></dd>

    <dt><code>kmer_length</code></dt>
<dd><p>a vector of kmer lengths to screen in order to find the best seed motif.</p></dd>

    <dt><code>max_cands</code></dt>
<dd><p>maximum number of kmer candidates to try.</p></dd>

    <dt><code>smooth_k</code></dt>
<dd><p>k for smoothing the predictions of each model in order to compute the residuals when <code>motif_num</code> &gt; 1. The residuals are computed as <code>response</code> - running mean of size 'k' of the current model.</p></dd>

    <dt><code>min_kmer_cor</code></dt>
<dd><p>minimal correlation between the kmer and the response in order to use it as a seed.</p></dd>

    <dt><code>is_train</code></dt>
<dd><p>a boolean vector that determine which subset of sequences to use when screening</p></dd>

    <dt><code>min_gap,max_gap</code></dt>
<dd><p>the length of a gap to be considered in the pattern. Only one gap, of length min_gap:max_gap, is being used, and is located anywhere in the motif. Note that this greatly expand the search space (and increase multiple testing severely).</p></dd>

    <dt><code>min_cor</code></dt>
<dd><p>Only patterns for which the maximum correlation to one of the response variable is larger than min_cor will be reported</p></dd>

    <dt><code>min_n</code></dt>
<dd><p>Only patterns for which the average number of occurrences in the sequences is larger than min_n will be reported</p></dd>

    <dt><code>from_range</code></dt>
<dd><p>Sequences will be considered only from position from_range (default 0)</p></dd>

    <dt><code>to_range</code></dt>
<dd><p>Sequences will be considered only up to position to_range (default NULL - using the length of the sequences)</p></dd>

    <dt><code>return_mat</code></dt>
<dd><p>Return a matrix of patterns and their correlation to the response variables instead of a data frame. (default: FALSE)</p></dd>

  
</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a list with the following elements:</p><ul><li><p>pssm: data frame with the pssm matrix with the inferred motif, where rows are positions and columns are nucleotides.</p></li>
<li><p>spat: a data frame with the inferred spatial model, with the spatial factor for each bin. The bins are defined such that the first bin starts at <code>spat_min</code> and the last bin ends at <code>spat_max</code>, with a bin size of <code>spat_bin</code>.</p></li>
<li><p>pred: a vector with the predicted pwm for each sequence.</p></li>
<li><p>consensus: Consensus sequence based on the PSSM.</p></li>
<li><p>response: The response matrix. If <code>include_response</code> is FALSE, the response matrix is not included in the list.</p></li>
<li><p>r2: \(r^2\) of the prediction with respect to the each response variable.</p></li>
<li><p>ks: If response is binary, Kolmogorov-Smirnov test results of the predictions where the response was 1 vs the predictions where the response was 0.</p></li>
<li><p>seed_motif: The seed motif that started the regression.</p></li>
<li><p>kmers: The k-mers that were screened in order to find the best seed motif (if motif was NULL).</p></li>
<li><p>sample_idxs: The indices of the sequences that were used for the regression (only for <code>regress_pwm.sample</code>).</p></li>
</ul><p>When <code>match_with_db</code> is TRUE, the following additional elements are returned:</p><ul><li><p>motif_db: The motif database that the most similar to the resulting PSSM.</p></li>
<li><p>db_match_dist: The distance between the resulting PSSM and the closest match in the motif database (KL divergence).</p></li>
<li><p>db_match_pssm: The PSSM of the closest match in the motif database.</p></li>
<li><p>db_match_pred: The predicted PWM of the closest match in the motif database.</p></li>
<li><p>db_match_r2: The \(r^2\) of the predicted PWM of the closest match in the motif database and the response</p></li>
<li><p>db_match_ks: If response is binary, the Kolmogorov-Smirnov test results of the predicted PWM of the closest match in the motif database where the response was 1 vs the predictions where the response was 0.</p></li>
</ul><p>When <code>screen_db</code> is TRUE, the following additional elements are returned:</p><ul><li><p>db_motif: The best motif from the motif database.</p></li>
<li><p>db_motif_pred: The predicted PWM of the best motif from the motif database.</p></li>
<li><p>db_motif_pssm: The PSSM of the best motif from the motif database.</p></li>
<li><p>db_motif_score: The score of the best motif from the motif database.</p></li>
</ul><p>When <code>n_motifs</code> is greater than 1, the following additional elements are returned:</p><ul><li><p>models: A list (as above) of each inferred model</p></li>
<li><p>multi_stats: A data frame with the following columns: <code>model</code>, <code>score</code> (KS for binary, r^2 otherwise), <code>comb_score</code> (score for the combined linear model for models 1:i) and additional statistics per model</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">regress_pwm.sample</span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">cluster_mat_example</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, final_metric <span class="op">=</span> <span class="st">"ks"</span>, screen_db <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">pssm</span></span></span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">spat</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pred</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Aviezer Lifshitz, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer></div>

  

  

  </body></html>

