<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Perform a PWM regression"><title>Perform a PWM regression — regress_pwm • prego</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Perform a PWM regression — regress_pwm"><meta property="og:description" content="Perform a PWM regression"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">prego</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.0.9000</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Perform a PWM regression</h1>
      
      <div class="d-none name"><code>regress_pwm.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Perform a PWM regression</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">regress_pwm</span><span class="op">(</span></span>
<span>  <span class="va">sequences</span>,</span>
<span>  <span class="va">response</span>,</span>
<span>  motif <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  motif_length <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  score_metric <span class="op">=</span> <span class="st">"r2"</span>,</span>
<span>  bidirect <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  spat_min <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  spat_max <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  spat_bin <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  improve_epsilon <span class="op">=</span> <span class="fl">0.0001</span>,</span>
<span>  min_nuc_prob <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  unif_prior <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  is_train <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  include_response <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">60427</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  kmer_length <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  motif_num <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>sequences</dt>
<dd><p>A vector of DNA sequences ('A', 'T', 'C' or 'G'. Will go through <code>toupper</code>)</p></dd>


<dt>response</dt>
<dd><p>A matrix of response variables - number of rows should equal the number of sequences</p></dd>


<dt>motif</dt>
<dd><p>Initial motif to start the regression from. Can be either a string with a kmer where the character "*" indicates a
wildcard or a data frame with a pre-computed PSSM (see thre slot <code>pssm</code> in the return value of this function).
If NULL - a K-mer screen would be performed in order to find the best kmer for initialization.</p></dd>


<dt>motif_length</dt>
<dd><p>Length of the seed motif. If the motif is shorter than this, it will be extended by wildcards (stars). Note that If the motif is longer than this, it will <em>not</em> be truncated.</p></dd>


<dt>score_metric</dt>
<dd><p>metric to use for optimizing the PWM. One of "r2" or "ks". When using "ks" the response variable should be a single vector of 0 and 1.</p></dd>


<dt>bidirect</dt>
<dd><p>is the motif bi-directional. If TRUE, the reverse-complement of the motif will be used as well.</p></dd>


<dt>spat_min</dt>
<dd><p>start of the spatial model from the beginning of the sequence (in bp)</p></dd>


<dt>spat_max</dt>
<dd><p>end of the spatial model from the beginning of the sequence (in bp). If NULL - the spatial model
would end at the end of the sequence.</p></dd>


<dt>spat_bin</dt>
<dd><p>size of the spatial bin (in bp).</p></dd>


<dt>improve_epsilon</dt>
<dd><p>minimum improve in the objective function to continue the optimization</p></dd>


<dt>min_nuc_prob</dt>
<dd><p>minimum nucleotide probability in every iteration</p></dd>


<dt>unif_prior</dt>
<dd><p>uniform prior for nucleotide probabilities</p></dd>


<dt>is_train</dt>
<dd><p>a boolean vector that determine which subset of sequences to use when screening</p></dd>


<dt>include_response</dt>
<dd><p>include the response in the resulting list (default: TRUE)</p></dd>


<dt>seed</dt>
<dd><p>random seed</p></dd>


<dt>verbose</dt>
<dd><p>show verbose messages.</p></dd>


<dt>kmer_length</dt>
<dd><p>The number of non-gap characters in motifs that will be screened</p></dd>


<dt>...</dt>
<dd><p>Arguments passed on to <code><a href="screen_kmers.html">screen_kmers</a></code></p><dl><dt><code>min_cor</code></dt>
<dd><p>Only patterns for which the maximum correlation to one of the response variable is larger than min_cor will be reported</p></dd>

    <dt><code>min_n</code></dt>
<dd><p>Only patterns for which the average number of occurrences in the sequences is larger than min_n will be reported</p></dd>

    <dt><code>min_gap,max_gap</code></dt>
<dd><p>the length of a gap to be considered in the pattern. Only one gap, of length min_gap:max_gap, is being used, and is located anywhere in the motif. Note that this greatly expand the search space (and increase multiple tesing severly).</p></dd>

    <dt><code>from_range</code></dt>
<dd><p>Sequences will be considered only from position from_range (default 0)</p></dd>

    <dt><code>to_range</code></dt>
<dd><p>Sequences will be considered only up to position to_range (default NULL - using the length of the sequences)</p></dd>

    <dt><code>return_mat</code></dt>
<dd><p>Return a matrix of patterns and their correlation to the response variables instead of a data frame. (default: FALSE)</p></dd>

  
</dl></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a list with the following elements:</p><ul><li><p>pssm: data frame with the pssm matrix with the inferred motif, where rows are positions and columns are nucleotides.</p></li>
<li><p>spat: a data frame with the inferred spatial model, with the spatial factor for each bin.</p></li>
<li><p>pred: a vector with the predicted pwm for each sequence.</p></li>
<li><p>response: The response matrix. If <code>include_response</code> is FALSE, the response matrix is not included in the list.</p></li>
<li><p>r2: \(r^2\) of the prediction with respect to the each response variable.</p></li>
<li><p>ks: If response is binary, Kolmogorov-Smirnov test results of the predictions where the response was 1 vs the predictions where the response was 0.</p></li>
<li><p>seed_motif: The seed motif that started the regression.</p></li>
<li><p>kmers: The k-mers that were screened in order to find the best seed motif (if motif was NULL).</p></li>
</ul></div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">regress_pwm</span><span class="op">(</span><span class="va">sequences_example</span>, <span class="va">response_mat_example</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Number of response variables: <span style="color: #0000BB;">5</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Screening for kmers in order to initialize regression</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Number of response variables: <span style="color: #0000BB;">5</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Screening kmers of length <span style="color: #0000BB;">8</span>, from position <span style="color: #0000BB;">0</span> to position <span style="color: #0000BB;">300</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> minimal correlation: <span style="color: #0000BB;">0.08</span>, minimal number of occurrences: <span style="color: #0000BB;">50</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> done normalizing response </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> new best CCACCCCC  0.00718373</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> new best CCCCACCC  0.0146177</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> done screening </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Found <span style="color: #0000BB;">2138</span> kmers in <span style="color: #0000BB;">1000</span> sequences.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Initializing regression with <span style="color: #0000BB;">"TTTACAAC"</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Running regression</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BB00;">✔</span> Finished running regression</span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">pssm</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   psid pos          A          C         G          T</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    0   0 0.17101707 0.31266135 0.1941226 0.32219896</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    0   1 0.20016445 0.28261185 0.1761953 0.34102842</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    0   2 0.26815110 0.24526116 0.3299702 0.15661749</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    0   3 0.80165291 0.04958678 0.0537190 0.09504131</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    0   4 0.03046471 0.42433441 0.2698603 0.27534056</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    0   5 0.37244797 0.19862147 0.2461770 0.18275347</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7    0   6 0.37228289 0.08648261 0.2420855 0.29914907</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 8    0   7 0.25095624 0.19095579 0.2256914 0.33239663</span>
<span class="r-in"><span><span class="va">res</span><span class="op">$</span><span class="va">spat</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   psid bin spat_factor</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1    0   0   0.1220389</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2    0  50   0.1319409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3    0 100   0.1985634</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4    0 150   0.1516449</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5    0 200   0.1319409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6    0 250   0.1319409</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 7    0 300   0.1319409</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">pred</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] -7.270413 -7.297840 -7.026858 -7.599799 -7.084934 -7.561258</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>`guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; = "none")` instead.</span>
<span class="r-plt img"><img src="regress_pwm-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># intialize with a pre-computed PSSM</span></span></span>
<span class="r-in"><span><span class="va">res1</span> <span class="op">&lt;-</span> <span class="fu">regress_pwm</span><span class="op">(</span><span class="va">sequences_example</span>, <span class="va">response_mat_example</span>, motif <span class="op">=</span> <span class="va">res</span><span class="op">$</span><span class="va">pssm</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #00BBBB;">ℹ</span> Number of response variables: <span style="color: #0000BB;">5</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>argument is not an atomic vector; coercing</span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in if (stringr::str_length(motif) &lt; motif_length) {    motif &lt;- stringr::str_pad(motif, motif_length, "*", side = "both")    cli_alert_info("Motif is shorter than {.val {motif_length}}, extending with wildcards")}:</span> the condition has length &gt; 1</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># binary response</span></span></span>
<span class="r-in"><span><span class="va">res_binary</span> <span class="op">&lt;-</span> <span class="fu">regress_pwm</span><span class="op">(</span><span class="va">cluster_sequences_example</span>, <span class="va">cluster_mat_example</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in regress_pwm(cluster_sequences_example, cluster_mat_example[,     1]):</span> object 'cluster_sequences_example' not found</span>
<span class="r-in"><span><span class="fu"><a href="plot_regression_qc.html">plot_regression_qc</a></span><span class="op">(</span><span class="va">res_binary</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error in dplyr::select(., -1):</span> object 'res_binary' not found</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Aviezer Lifshitz, Amos Tanay.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

    </footer></div>

  

  

  </body></html>

